hydromt/data_sources/data_source.py

```diff
@@ -0,0 +1,53 @@
+"""Abstract DataSource class."""
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from hydromt import DataCatalog
+from hydromt.metadata_resolvers.resolver_plugin import RESOLVERS
+
+
+class DataSource(BaseModel):
+    """
+    A DataSource is a parsed section of a DataCatalog.
+
+    The DataSource, specific for a data type within HydroMT, is responsible for
+    validating the input from the DataCatalog, to
+    ensure the workflow fails as early as possible. A DataSource has information on
+    the driver that the data should be read with, and is responsible for initializing
+    this driver.
+    """
+
+    @classmethod
+    def from_catalog(
+        cls,
+        catalog: DataCatalog,
+        key: str,
+        provider: str | None = None,
+        version: str | None = None,
+    ) -> "DataSource":
+        """Create Data source from DataCatalog."""
+        return cls.model_validate(
+            catalog.get_source(key, provider, version).update("name", key)
+        )
+
+    def _validate_metadata_resolver(cls, v: Any):
+        assert isinstance(v, str), "metadata_resolver should be string."
```

Comment:
Not sure how common this is, but asserts can get turned off when someone runs python with specific settings iirc which means that under certain circumstances these checks would dissapear. If these are critical we should just use if -raise statements even though that's more laborious